catkin_ws/src/state_estimation/src/sensors.py

```diff
@@ -3,309 +3,165 @@
+DEG_PER_RAD = 180 / np.pi
+RAD_PER_DEG = 1 / DEG_PER_RAD
 
-# global frame relative to NED (North-East-Down)
-Q_GLOBAL_NED = np.quaternion(0, 1, 0, 0) # inertial frame, will not change
-TIME_UNTIL_INACTIVE = 1 # after 1s of no data, consider sensor is inactive # TODO - set
-
-# TODO - instead of strat by sensor, strat by data?
-class Sensor:
-    def __init__(self):
-        self.is_active = False
-        self.last_unique_msg_time = None
-        self.last_unique_msg = None
-        rospy.Timer(rospy.Duration(0.1), self.check_active) #TODO - duration, needs to be shutdown?
-
-
-    def q_auv_global(self):
-        return None
-
-
-    def pos_auv_global(self, q_auv_global):
-        return None
 
+class Sensor():
+    def __init__(self, sensor_name):
+        self.time_before_considered_inactive = 1 #seconds
+        self.sensor_name = sensor_name
 
-    # TODO - handle case when sensor publisher latches on last message when sensor inactive
-    def check_active(self, _): # TODO - what is second arg?
-        # if inactive - remain that way
-        # if active - may have become inactive due to no recent messages
-        if self.is_active and rospy.get_time() - self.last_unique_msg_time > TIME_UNTIL_INACTIVE:
-            self.is_active = False
-            rospy.logwarn("state_estimation sensors - " + self.id + " has become inactive")
-            # TODO - update frames so that using other sensors coincides with current state_estimates
-
+        self.x = 0
+        self.y = 0
+        self.z = 0
+        self.roll = 0
+        self.pitch = 0
+        self.yaw = 0
+        
+        self.q_nwu_auv = np.quaternion(1, 0, 0, 0)
+        self.angular_velocity = np.array([0,0,0])
+
+        # initialize a sensor as "inactive"
+        self.last_unique_state_time = 0
+        self.last_state = [self.x,self.y,self.z,self.roll,self.pitch,self.yaw,self.q_nwu_auv,self.angular_velocity]
+    
+    def updateLastState(self):
+        current_state = [self.x,self.y,self.z,self.roll,self.pitch,self.yaw,self.q_nwu_auv,self.angular_velocity]
+        for i in range(len(current_state)):
+            if isinstance(current_state[i], Iterable): 
+                different = any([vc != vl for vc, vl in zip(current_state[i], self.last_state[i])])
+            else:
+                different = current_state[i] != self.last_state[i]
+            if different:
+                if rospy.get_time() - self.last_unique_state_time > self.time_before_considered_inactive:
+                    rospy.loginfo("{} has become active.".format(self.sensor_name))
+                    self.last_unique_state_time = rospy.get_time() 
+                    break
+        self.last_state = current_state
+        
+    def isActive(self):
+        if rospy.get_time() - self.last_unique_state_time > self.time_before_considered_inactive:
+            rospy.logwarn("{} has been inactive for {} seconds.".format(self.sensor_name, self.time_before_considered_inactive))
```

Comment:
need to fix this (maybe its already done) - the status error message spams the console its really annoying

Reply:
fixed